hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -949,17 +935,81 @@ private TcpClientConnection onAuthenticated(TcpClientConnection connection,
         return connection;
     }
 
-    private void checkClientStateOnClusterIdChange(TcpClientConnection connection) {
+    /**
+     * Checks the client state against the intend of the callee(switchingToNextCluster)
+     * closes the connection and throws exception if the authentication needs to be cancelled.
+     */
+    private void checkClientState(TcpClientConnection connection, boolean switchingToNextCluster) {
+        if (clientState == ClientState.SWITCHING_CLUSTER && !switchingToNextCluster) {
+            String reason = "There is a cluster switch in progress. "
+                    + "This connection attempt initiated before the progress and not allowed to be authenticated.";
+            connection.close(reason, null);
+            throw new AuthenticationException(reason);
+        }
+        //Following state can not happen. There is only one path with `switchingToNextCluster` as true
+        //and that path starts only when the old switch fails. There are no concurrent run of that path.
+        if (clientState != ClientState.SWITCHING_CLUSTER && switchingToNextCluster) {
```

Comment:
Am I missing something in the following scenario?

We have 2 clusters (A and B), we are connected to A. We are at the `INITIALIZED_ON_CLUSTER` state. Then, A goes down and does not come back. Client triggers reconnection (via `triggerClusterReconnection`), we first try to connect A (current cluster), connection attempts fail (since A is down), then we will try to connect B. Now, we will set the `switchingToNextCluster` flag to true, and let's imagine that we authenticated on B. We will come here, and close the connection since the client is at the `INITIALIZED_ON_CLUSTER` state and `switchingToNextCluster` is true.

Reply:
Good catch.  I have send a second commit for to fix it. 
And I have tested the related enterprise tests. They are passing now.

Reply:
I feel like there is still a problem. Imagine that we have provided 2 cluster configs. Assume that the first one is down, the second one is up. We will try to connect to the first one, which will fail. The state will still be `INITIAL`. Now, we will try the second one, will set the `switchingToNextCluster` to True on `destroyCurrentClusterConnectionAndTryNextCluster` and come here directly (note that we will skip `checkClientStateOnClusterIdChange` since the cluster id is not changed yet because it was null in the first place) and again close the connection

Reply:
I have sent one more commit. I have also included this scenario as a test on the enterprise repo. hazelcast/hazelcast-enterprise#4029